### PR TITLE
Make text array type GPDB hashable.

### PR DESCRIPTION
**Pull request by guofengrichard**
By design, all array types should be hashable by GPDB.
Issue #3741 shows that GPDB suffers from text array type
being not hashable error.

This commit fixes that.

### CODE REVIEW

File: src/backend/executor/nodeResult.c [hunk 1]
```diff
@@ -289,19 +289,9 @@ static bool TupleMatchesHashFilter(Result *resultNode, TupleTableSlot *resultSlo
...
-					case INT2ARRAYOID:
-					case INT4ARRAYOID:
-					case INT8ARRAYOID:
-					case FLOAT4ARRAYOID:
-					case FLOAT8ARRAYOID:
-					case REGTYPEARRAYOID:
-						att_type = ANYARRAYOID;
-						/* fall through */
-					default:
-						break;
```
Comment (vraghavan78) [inline src/backend/executor/nodeResult.c]: shouldn't we also change bool isGreenplumDbHashable(Oid typid)?

**Comment by d** [comment]
> By design

This is a strong assertion, @guofengrichard . Can you point us to the evidence of that "design"?

**Comment by guofengrichard** [comment]
Thanks @danielgustafsson for the review. @d As Daniel pointed out, several code places are showing  that GPDB can hash all array types.

**Comment by d** [comment]
I regret not giving this a -1, my bad. :cry: 

## TL;DR
To be clear, I still believe #3741 is worth fixing, but not like this (#4940).

## Long version
Arrays that are hashable are exception, not the norm: we hash the whole array as a buffer, which means that elements with non-unique representation (equality-wise) will likely have different hash values. For example, `int[]` and `oidvector` are hashable arrays, `text[]` are also hashable. But `float[]` are not:

```sql
CREATE TEMP TABLE foo(a int, b float[]) DISTRIBUTED BY (b);
INSERT INTO foo VALUES (1, '{-0, -0}'::float[]), (1, '{-0, 0}'::float[]);

CREATE TEMP TABLE bar(c int, d float[]) DISTRIBUTED BY (c);
INSERT INTO bar(c,d) SELECT a, b FROM foo;
SELECT gp_segment_id, b, a FROM foo;
set optimizer to off;
SELECT distinct(b) FROM foo;
SELECT distinct(d) FROM bar;
```

This is the output I got (with 3 segments):
```
CREATE TEMP TABLE foo(a int, b float[]) DISTRIBUTED BY (b);
CREATE TABLE
INSERT INTO foo VALUES (1, '{-0, -0}'::float[]), (1, '{-0, 0}'::float[]);
INSERT 0 2
CREATE TEMP TABLE bar(c int, d float[]) DISTRIBUTED BY (c);
CREATE TABLE
INSERT INTO bar(c,d) SELECT a, b FROM foo;
INSERT 0 2
SELECT gp_segment_id, b, a FROM foo;
 gp_segment_id |    b    | a 
---------------+---------+---
             1 | {-0,-0} | 1
             0 | {-0,0}  | 1
(2 rows)

set optimizer to off;
SET
SELECT distinct(b) FROM foo;
    b    
---------
 {-0,-0}
 {-0,0}
(2 rows)

SELECT distinct(d) FROM bar;
   d    
--------
 {-0,0}
(1 row)
```

**Comment by d** [comment]
Ah this was the case I failed to produce, I was thinking of how I can get an array with a null bitmap but with none of the bits set…
